services/inc/ref_count.h

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (c) 2023 Particle Industries, Inc.  All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation, either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <type_traits>
+#include <utility>
+#include <atomic>
+
+static_assert(std::atomic_int::is_always_lock_free, "std::atomic<int> is not always lock-free");
+
+namespace particle {
+
+/**
+ * Base class for reference counted objects.
+ */
+class RefCount {
+public:
+    RefCount() :
+            count_(1) {
+    }
+
+    RefCount(const RefCount&) = delete;
+
+    virtual ~RefCount() = default;
+
+    void addRef() const {
+        count_.fetch_add(1, std::memory_order_relaxed);
+    }
+
+    void release() const {
+        if (count_.fetch_sub(1, std::memory_order_acq_rel) == 1) {
+            delete this;
+        }
+    }
+
+    RefCount& operator=(const RefCount&) = delete;
+
+private:
+    mutable std::atomic_int count_;
+};
+
+/**
+ * Smart pointer for reference counted objects.
+ */
+template<typename T>
+class RefCountPtr {
+public:
+    RefCountPtr() :
+            RefCountPtr(nullptr) {
+    }
+
+    RefCountPtr(T* ptr) :
```

Comment:
I get the semantics of this constructor and the copy constructor incrementing the reference count, and the destructor decreasing it. I don't understand the semantics of `wrap` and `unwrap` and why this class needs both interfaces.

Reply:
Looking at more of the code, `wrap` and `unwrap` are used to convert between the the smart pointer and the raw pointer type, sometimes when passing the pointer to a container that deals with opaque types. That makes sense, but the usage is hard to follow because many uses of `unwrap()` are paired with a call to `release()` on the reference counted object. Ideally, every use of `wrap()` would be paired with a user of `unwrap()` and there would be no need to call `release()` directly the same way there is no need to call `addRef()` directly.

^^ The above paragraph may not be 100% accurate, but it's a note that it's hard to reason about the ownership and safety of the smart pointers above. Usually I'd look for matching pairs calls to `new/delete` or `malloc/free` in add/remove functions. Here's it's harder to do this.

Reply:
Julien and I discussed this at length. My main takeaway from the feedback is that `release()` and `addRef()` are considered a low-level API that should not ideally be used in C++ code for which there's a smart pointer class `RefCountPtr`. I agree with that, however, I think there's at least one exception where direct usage of `release()` can be justified: if there's a reference to an object that is _not managed_ by a smart pointer in the first place, it's should be okay to call `release()` on it as it would, arguably, express the intent in a more direct and readable way as opposed to wrapping the pointer via `RefCountPtr::wrap()` and letting the resulting smart pointer immediately go out of scope.

In the code in question, `release()` is only called on objects that either came through the C-based dynalib interface or are stored in an intrusive list container where it's not possible to use smart pointers. For those objects there's always a "dangling" reference and `release()` destroys it in an explicit way.

With that said, this is quite a subjective topic. I'd be okay replacing all occurrences of `release()` with calls to `wrap()` if we find the latter more readable.

Reply:
I'm OK with keeping the usage of `release()` in those cases. I'd prefer additional comments around the ownership changes because of the bad consequences of getting it wrong for future maintainers.

Reply:
Added more comments for now.